vms/platformvm/network/warp.go

```diff
@@ -0,0 +1,336 @@
+			Message: fmt.Sprintf("subnet %q has not been converted", subnetID),
+		}
+	}
+	if err != nil {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to get subnet conversionID: " + err.Error(),
+		}
+	}
+
+	if msg.ID != conversion.ConversionID {
+		return &common.AppError{
+			Code:    ErrMismatchedConversionID,
+			Message: fmt.Sprintf("provided conversionID %q != expected conversionID %q", msg.ID, conversion.ConversionID),
+		}
+	}
+
+	return nil
+}
+
+func (s signatureRequestVerifier) verifySubnetValidatorRegistration(
+	msg *message.SubnetValidatorRegistration,
+	justificationBytes []byte,
+) *common.AppError {
+	if msg.Registered {
+		return s.verifySubnetValidatorRegistered(msg.ValidationID)
+	}
+
+	var justification platformvm.SubnetValidatorRegistrationJustification
+	if err := proto.Unmarshal(justificationBytes, &justification); err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseJustification,
+			Message: "failed to parse justification: " + err.Error(),
+		}
+	}
+
+	switch preimage := justification.GetPreimage().(type) {
+	case *platformvm.SubnetValidatorRegistrationJustification_ConvertSubnetTxData:
+		return s.verifySubnetValidatorNotCurrentlyRegistered(msg.ValidationID, preimage.ConvertSubnetTxData)
+	case *platformvm.SubnetValidatorRegistrationJustification_RegisterSubnetValidatorMessage:
+		return s.verifySubnetValidatorCanNotValidate(msg.ValidationID, preimage.RegisterSubnetValidatorMessage)
+	default:
+		return &common.AppError{
+			Code:    ErrInvalidJustificationType,
+			Message: fmt.Sprintf("invalid justification type: %T", justification.Preimage),
+		}
+	}
+}
+
+// verifySubnetValidatorCanNotValidate verifies that the validationID is
+// currently a validator.
+func (s signatureRequestVerifier) verifySubnetValidatorRegistered(
+	validationID ids.ID,
+) *common.AppError {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	// Verify that the validator exists
+	_, err := s.state.GetSubnetOnlyValidator(validationID)
+	if err == database.ErrNotFound {
+		return &common.AppError{
+			Code:    ErrValidationDoesNotExist,
+			Message: fmt.Sprintf("validation %q does not exist", validationID),
+		}
+	}
+	if err != nil {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to get subnet only validator: " + err.Error(),
+		}
+	}
+	return nil
+}
+
+// verifySubnetValidatorCanNotValidate verifies that the validationID is not
+// currently a validator.
+func (s signatureRequestVerifier) verifySubnetValidatorNotCurrentlyRegistered(
+	validationID ids.ID,
+	justification *platformvm.SubnetIDIndex,
+) *common.AppError {
+	subnetID, err := ids.ToID(justification.GetSubnetId())
+	if err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseSubnetID,
+			Message: "failed to parse subnetID: " + err.Error(),
+		}
+	}
+
+	justificationID := subnetID.Append(justification.GetIndex())
+	if validationID != justificationID {
```

Comment:
The `mismatched convert subnet validationID` test shows why this case is needed.

If this check were not here, then it would be possible to get nodes to sign a warp message saying that a node will not be registered to the validator set, even though it could be.